False positive `used-before-assignment` in pattern matching with a guard.

### Bug description

When using pattern matching with a guard, `pylint` claims that the variable the guard uses is used before assignment. 

### Example:

```python
match ("example", "one"):
    case (x, y) if x == "example":  # <-- pylint complains about 'x' here
        print("woo!")
    case _:
        print("boo...")
```


### Configuration

_No response_

### Command used

```shell
pylint a.py
```


### Pylint output

```shell
************* Module a
a.py:2:19: E0601: Using variable 'x' before assignment (used-before-assignment)

-------------------------------------
Your code has been rated at -10.00/10
```


### Expected behavior

Pattern matching with guards is correctly parsed and supported.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]
```


### OS / Environment

Mac OSX Big Sur 11.6 (20G165)

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/5327.false_positive
+++ b/doc/whatsnew/fragments/5327.false_positive
@@ -1,0 +1,4 @@
+Fix false-positive for ``used-before-assignment`` in pattern matching
+with a guard.
+
+Closes #5327

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2139,6 +2139,7 @@ class VariablesChecker(BaseChecker):
                             nodes.AugAssign,
                             nodes.Expr,
                             nodes.Return,
+                            nodes.Match,
                         ),
                     )
                     and VariablesChecker._maybe_used_and_assigned_at_once(defstmt)
@@ -2239,6 +2240,8 @@ class VariablesChecker(BaseChecker):
         """Check if `defstmt` has the potential to use and assign a name in the
         same statement.
         """
+        if isinstance(defstmt, nodes.Match):
+            return any(case.guard for case in defstmt.cases)
         if isinstance(defstmt.value, nodes.BaseContainer) and defstmt.value.elts:
             # The assignment must happen as part of the first element
             # e.g. "assert (x:= True), x"

--- a/tests/functional/u/used/used_before_assignment_py310.py
+++ b/tests/functional/u/used/used_before_assignment_py310.py
@@ -1,0 +1,7 @@
+"""Tests for used-before-assignment with python 3.10's pattern matching"""
+
+match ("example", "one"):
+    case (x, y) if x == "example":
+        print("x used to cause used-before-assignment!")
+    case _:
+        print("good thing it doesn't now!")

--- a/tests/functional/u/used/used_before_assignment_py310.rc
+++ b/tests/functional/u/used/used_before_assignment_py310.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.10

